File: java/cpp/dataStructures/graphs/Digraph.cpp
/** @file Digraph.cpp
 *
 *  @author Jon Turner
 *  @date 2011
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

#include "stdinc.h"
#include "Digraph.h"

/** Construct a Digraph with space for a specified number of vertices and edges.
 *  @param numv is the maximum number of vertices in the graph
 *  @param maxe is the maximum number of edges
 */
Digraph::Digraph(int numv, int maxe) : Graph(numv,maxe) {
	makeSpace(numv,maxe);
}

/** Copy constructor - not implemented. */
Digraph::Digraph(const Digraph& original) : Graph(original) { }

/** Allocate and initialize dynamic storage for graph.
 */
void Digraph::makeSpace(int numv, edge maxe) {
	fi = new edge[numv+1];
	for (vertex u = 0; u <= numv; u++) fi[u] = 0;
} 

/** Free space used by Digraph */
Digraph::~Digraph() { freeSpace(); }

void Digraph::freeSpace() {
	delete [] fi;
}

/** Copy another Graph to this one.
 *  @param original is another graph that is to replace this one.
 */
void Digraph::copyFrom(const Digraph& original) {
        if (N != original.n() || maxEdge < original.m()) {
                resize(original.n(), original.m());
        } else {
                reset();
        }
        N = original.n();
        for (edge e = original.first(); e != 0; e = original.next(e)) {
                join(original.left(e),original.right(e));
        }
        sortAdjLists();
}

/** Discard current graph and allocate space for a new graph.
 *  @param numv is the number of vertices in the new graph
 *  @param maxe is the maximum number of edges in the new graph
 */
void Digraph::resize(int numv, int maxe) {
        Graph::resize(numv, maxe);
        freeSpace(); makeSpace(numv, maxe);
}

/** Remove all edges from the graph.  */
void Digraph::reset() {
	Graph::reset();
	for (vertex u = 0; u <= n(); u++) fi[u] = 0;
}

/** Join two vertices with an edge.
 *  @param u is the tail of the new edge
 *  @param v is the head of the new edge
 *  @return the edge number for the new edge, or 0 on failure
 */
edge Digraph::join(vertex u, vertex v) { return Graph::join(u,v); }

/** Join two vertices with a specific edge.
 *  @param u is the tail of the new edge
 *  @param v is the head of the new edge
 *  @param e is the number of an idle edge
 *  @return the edge number for the new edge, or 0 on failure
 */
edge Digraph::joinWith(vertex u, vertex v, edge e) {
	assert(validVertex(u) && validVertex(v));

	if (e == 0 || !edges->isOut(e)) return 0;
	edges->swap(e);

	// initialize edge information
	evec[e].l = u; evec[e].r = v;

	// add edge to the adjacency lists
	// in the adjLists data structure, each edge appears twice,
	// as 2*e and 2*e+1
	if (fe[u] != 0) {
		adjLists->join(2*e,fe[u]);
	}
	if (fi[v] != 0) {
		adjLists->join(2*e+1,fi[v]);
	}
	if (fe[u] == 0) fe[u] = 2*e;
	if (fi[v] == 0) fi[v] = 2*e+1;

	M++;

	return e;
}

string& Digraph::edge2string(edge e, string& s) const {
	return Graph::edge2string(e,s);
}

/** Create a string representation of an adjacency list.
 *  @param u is a vertex number
 *  @param s is a reference to a string in which the result is returned
 *  @return a reference to s.
 */
string& Digraph::alist2string(vertex u, string& s) const {
        s = "";
        if (firstAt(u) == 0) return s;
        int cnt = 0;
        for (edge e = firstOut(u); e != 0; e = nextOut(u,e)) {
                string s1;
                s += edge2string(e,s1) + " ";
                cnt++;
                if (cnt >= 10) { s += "\n"; cnt = 0; }
        }
        if (cnt != 0) s += "\n";
        return s;
}

/** Construct a string in dot file format representation 
 * of the Directed Graph object.
 *  For small graphs (at most 26 vertices), vertices are
 *  represented in the string as lower case letters.
 *  For larger graphs, vertices are represented by integers.
 *  @param s is a string object provided by the caller which
 *  is modified to provide a representation of the Graph.
 *  @return a reference to the string
 */
string& Digraph::toDotString(string& s) const {
	stringstream ss;
	// directed graph
	ss << "digraph G { " << endl;
	for (vertex u = 1; u <= n(); u++) {
		if (firstAt(u) == 0) break;
		string su;
		for (edge e = firstOut(u); e != 0; e = nextOut(u,e)) {
                        string s1, s2;
			vertex v = mate(u,e);
			su += Util::node2string(u,n(),s1) + " -> ";
			su += Util::node2string(v,n(),s2);
			su += " ; "; 
		}
                if (!su.empty())   ss << su << endl;
	}
	ss << " } " << endl;
	s = ss.str();
	return s;
}

/** Read one edge from an input stream, add it to the graph.
 *  @param in is an open input stream
 *  @return true on success, false on error.
 */
bool Digraph::readEdge(istream& in) {
        vertex u, v;
        if (Util::readNext(in,'(') == 0 || !Util::readNode(in,u,n()) ||
            Util::readNext(in,',') == 0 || !Util::readNode(in,v,n()) ||
            Util::readNext(in,')') == 0) {
                return false;
        }
        join(u,v);
        return true;
}

/** Read a graph.
 *  @param in is an open input stream
 *  @return true on success, else false
 */
bool Digraph::read(istream& in) {
        int numv, maxe;
        in >> numv >> maxe;
	if (N != numv || maxEdge < maxe) resize(numv,maxe);
	else reset();
	N = numv;
        for (int i = 1; i <= maxe; i++) {
                if (!readEdge(in)) return false;
        }
        if (M != maxe) return false;
        sortAdjLists();
        return true;
}

/** Generate a random digraph.
 *  @param numv is the number of vertices on which the digraph is generated;
 *  if this object has N>numv, the random graph is defined over the first
 *  numv vertices, leaving the remaining vertices with no edges
 *  @param nume is the number of edges in the generated digraph
 */
void Digraph::rgraph(int numv, int nume) {
	numv = max(0,numv); nume = max(0,nume);
	if (numv > N || nume > maxEdge) resize(numv,nume); 
        else reset();

	// build set containing edges already in graph
	HashSet edgeSet(nume);
	for (edge e = first(); e != 0; e = next(e)) {
		vertex u = min(tail(e), head(e));
		vertex v = max(tail(e), head(e));
		uint64_t vpair = u; vpair <<= 32; vpair |= v;
		edgeSet.insert(vpair);
	}

	// add edges using random sampling of vertex pairs
	// stop early if graph gets so dense that most samples
	// repeat edges already in graph
	while (m() < nume && m()/numv < numv/2) {
		vertex u = randint(1,numv);
		vertex v = randint(1,numv);
		if (u == v) continue;
		uint64_t vpair = u; vpair <<= 32; vpair |= v;
		if (!edgeSet.member(vpair)) {
			edgeSet.insert(vpair); join(u,v);
		}
	}
	if (m() == nume) return;

	// if more edges needed, build a vector containing remaining 
	// "candidate" edges and then sample from this vector
	vector<uint64_t> vpVec;
	unsigned int i = 0;
	for (vertex u = 1; u < numv; u++) {
		for (vertex v = 1; v <= numv; v++) {
			if (v == u) continue;
			uint64_t vpair = u; vpair <<= 32; vpair |= v;
			if (!edgeSet.member(vpair)) vpVec[i++] = vpair;
		}
	}
	// sample remaining edges from vector
	i = 0;
	while (m() < nume && i < vpVec.size()) {
		int j = randint(i,vpVec.size()-1);
		vertex u = vpVec[j] >> 32;
		vertex v = vpVec[j] & 0xffffffff;
		join(u,v); vpVec[j] = vpVec[i++];
	}
	sortAdjLists();
}

/** Generate a random directed acyclic graph.
 *  @param numv is the number of vertices on which the digraph is generated;
 *  if this object has N>numv, the random graph is defined over the first
 *  numv vertices, leaving the remaining vertices with no edges
 *  @param nume is the number of edges in the generated digraph
 */
void Digraph::rdag(int numv, int nume) {
	numv = max(0,numv); nume = max(0,nume);
	if (N < numv || maxEdge < nume) resize(numv,nume); 
        else reset();

	// build set containing edges already in graph
	HashSet edgeSet(nume);
	for (edge e = first(); e != 0; e = next(e)) {
		vertex u = min(tail(e), head(e));
		vertex v = max(tail(e), head(e));
		uint64_t vpair = u; vpair <<= 32; vpair |= v;
		edgeSet.insert(vpair);
	}

	// add edges using random sampling of vertex pairs
	// stop early if graph gets so dense that most samples
	// repeat edges already in graph
	while (m() < nume && m()/numv < numv/4) {
		vertex u = randint(1,numv-1);
		vertex v = randint(u+1,numv);
		if (u == v) continue;
		uint64_t vpair = u; vpair <<= 32; vpair |= v;
		if (!edgeSet.member(vpair)) {
			edgeSet.insert(vpair); join(u,v);
		}
	}
	if (m() == nume) return;

	// if more edges needed, build a vector containing remaining 
	// "candidate" edges and then sample from this vector
	vector<uint64_t> vpVec;
	unsigned int i = 0;
	for (vertex u = 1; u < numv; u++) {
		for (vertex v = u+1; v <= numv; v++) {
			if (v == u) continue;
			uint64_t vpair = u; vpair <<= 32; vpair |= v;
			if (!edgeSet.member(vpair)) vpVec[i++] = vpair;
		}
	}
	// sample remaining edges from vector
	i = 0;
	while (m() < nume && i < vpVec.size()) {
		int j = randint(i,vpVec.size()-1);
		vertex u = vpVec[j] >> 32;
		vertex v = vpVec[j] & 0xffffffff;
		join(u,v); vpVec[j] = vpVec[i++];
	}
	sortAdjLists();
}

File: js/dataStructures/graphs/Digraph.mjs
/** @file Digraph.js
 *
 *  @author Jon Turner
 *  @date 2021
 *  This is open source software licensed under the Apache 2.0 license.
 *  See http://www.apache.org/licenses/LICENSE-2.0 for details.
 */

import { scramble, shuffle } from '../../common/Random.mjs';
import Scanner from '../basic/Scanner.mjs';
import Graph from './Graph.mjs';
import { randomPermutation } from '../../common/Random.mjs';

import { assert, EnableAssert as ea } from '../../common/Assert.mjs';

/** Data structure for directed graph.
 *  Extends Graph class and places incoming edges before outgoing edges
 *  in adjuacency lists.
 */
export default class Digraph extends Graph {
	firstEpOut		// firstEpOut[u] is endpoint of first outgoing edge
					// in u's adjacency list

	/** Constructor for directed graph
	 *  @param n is the number of vertices
	 *  @param erange is the max number of edges to provide space for
	 */
	constructor(n, erange=n) {
		super(...arguments);
		this.firstEpOut = new Int32Array(this.n+1);
	} 

	/** Get the tail of a directed edge.
	 *  @param e is a directed edge
	 *  @return the vertex the e leaves from
	 */ 
	tail(e) { return this.left(e); }

	/** Get the head of a directed edge.
	 *  @param e is a directed edge
	 *  @return the vertex the e goes to
	 */ 
	head(e) { return this.right(e); }

	/** Get the first edge coming into u.
	 *  @param u is a vertex
	 *  @return the first edge in the list of edges entering u, or 0 if
	 *  no such edge
	 */ 
	firstInto(u) {
		let e = this.firstAt(u);
		return ((e == 0 || u == this.tail(e)) ? 0 : e);
	}

	/** Get the next edge coming into u.
	 *  @param u is a vertex
	 *  @param e is an edge entering u
	 *  @return the next edge in the list of edges entering u, or 0 if
	 *  no such edge
	 */ 
	nextInto(u, e) {
		e = this.nextAt(u, e);
		return ((e == 0 || u == this.tail(e)) ? 0 : e);
	}

	/** Get the first edge leaving u.
	 *  @param u is a vertex
	 *  @return the first edge in the list of edges leaving u, or 0 if
	 *  no such edge
	 */ 
	firstOutof(u) { return Math.trunc(this.firstEpOut[u]/2); }

	/** Get the next edge leaving u.
	 *  @param u is a vertex
	 *  @param e is an edge leaving u
	 *  @return the next edge in the list of edges leaving u, or 0 if
	 *  no such edge
	 */ 
	nextOutof(u, e) { return this.nextAt(u, e); }

	/** Compute the in-degree of a vertex.
	 *  @param u is a vertex
	 *  @return the number of edges incident to u
	 */
	inDegree(u) {
		ea && assert(this.validVertex(u));
		let d = 0;
		for (let e = this.firstInto(u); e; e = this.nextInto(u,e)) {
			 d++;
		}
		return d;
	}

	/** Compute the out-degree of a vertex.
	 *  @param u is a vertex
	 *  @return the number of edges incident to u
	 */
	outDegree(u) {
		ea && assert(this.validVertex(u));
		let d = 0;
		for (let e = this.firstOutof(u); e; e = this.nextOutof(u,e)) d++;
		return d;
	}

	/** Find a directed edge joining two vertices
	 *  @param u is a vertex
	 *  @param v is another vertex
	 *  @return an edge from u to v, or 0 if no such edge
	 */
	findEdge(u, v, edges) {
		ea && assert(this.validVertex(u) && this.validVertex(v));
		if (!edges) {
			for (let e = this.firstOutof(u); e; e = this.nextOutof(u, e)) {
				if (v == this.mate(u, e)) return e;
			}
			return 0;
		}
		// do binary search in edges
		let lo = 0; let hi = edges.length-1;
		if (u > v) [u, v] = [v, u];
		while (lo < hi) {
			let mid = Math.floor((lo + hi) / 2);
			let e = edges[mid];
			let t = this.tail(e); let h = this.head(e);
				 if (u < t || (u == t && v < h)) hi = mid-1;
			else if (u > t || (u == t && v > h)) lo = mid+1;
			else return e;
		}
		return 0;
	}

	/** Join two vertices.
	 *  @param u is the tail for the new edge
	 *  @param v is the head of the new edge
	 *  @param e is the number of an "unused" edge;
	 *  if omitted, the first unused edge is used
	 *  @return the edge number for the new edge or 0
	 *  on failure
	 */
	join(u, v, e=this.edges.first(2)) {
		ea && assert(this.validVertex(u) && this.validVertex(v) &&
					 (e > 0 || this.edges.first(2) == 0));
/*
		if (u > this.n || v > this.n || this.edges.length(2) == 0) {
			this.expand(Math.max(this.n, u, v), Math.max(e, this.edges.n+1));
		}
*/
		if (e == 0) e = this.edges.first(2);
		this.edges.swap(e);

		// initialize edge information
		this.Left[e] = u; this.Right[e] = v;
	
		// add edge to the endpoint lists
		if (this.firstEpOut[u] == 0) this.firstEpOut[u] = 2*e;
			// this.firstEpOut[u] changes only with first outgoing edge
		this.firstEp[u] = this.epLists.join(this.firstEp[u], 2*e);
		this.firstEp[v] = this.epLists.join(2*e+1, this.firstEp[v]);
			// incoming edges inserted at the front of the list,
			// to keep them separate from outputs
	
		return e;
	}

	/** Delete an edge from the graph.
	 *  @param e is the edge to be deleted.
	 */
	delete(e) {
		ea && assert(this.validEdge(e));
		let u = this.left(e);
		if (e == this.firstOutof(u))
			this.firstEpOut[u] = 2 * this.nextOutof(u,e);
		super.delete(e);
	}

	/** Compare two edges incident to the same endpoint u.
	 *  @return -1 if u's mate in e1 is less than u's mate in e2,
	 *  return +1 if u's mate in e1 is greater than than u's mate in e2,
	 *  return  0 if u's mate in e1 is equal to its mate in e2.
	 */
	ecmp(e1, e2, u) {
		ea && assert(this.validVertex(u) && this.validEdge(e1)
									&& this.validEdge(e2));
			 if (u == this.head(e1) && u == this.tail(e2)) return -1;
		else if (u == this.tail(e1) && u == this.head(e2)) return 1;
		let v1 = this.mate(u, e1); let v2 = this.mate(u, e2);
		return (v1 != v2 ? v1 - v2 : this.compareProps(e1, e2));
	}

	sortEplist(u) {
		super.sortEplist(u);
		// now determine the firstEpOut value for re-ordered list
		for (let e = this.firstAt(u); e; e = this.nextAt(u,e)) {
				if (u == this.tail(e)) {
				this.firstEpOut[u] = 2*e; break;
			}
		}
	}

	/** Compute a sorted list of edge numbers.
	 *  @param evec is an optional array of edge numbers; if present,
	 *  its edges are sorted; if omitted, all edges in graph are sorted
	 *  @return a sorted array of edge numbers, where edges  are sorted
	 *  first by the tail, then by the head; if edge properties are present,
	 *  they are used to break ties.
	 */
    sortedElist(evec=null) {
        // first create vector of edge information
        // [tail, head, edge number]
        if (evec == null) {
			let i = 0; evec = new Array(this.m);
			for (let e = this.first(); e; e = this.next(e))
				evec[i++] = e;
		}
		evec.sort((e1,e2) => { let [t1,h1] = [this.tail(e1), this.head(e1)];
							   let [t2,h2] = [this.tail(e2), this.head(e2)];
							   return t1 != t2 ? t1-t2 :
									   h1 != h2 ? h1-h2 :
							   		  	this.compareProps(e1,e2);
							   });
		return evec;
	}
	
	/** Compare another graph to this one.
	 *  @param that is a Digraph object or a string representing one.
	 *  @return true if g is equal to this; that is, it has the same
	 *  vertices, edges and edge lengths. Properties named 'length"
	 *  are checked, but others are not.
	 */
	equals(that) {
		that = super.equals(...(arguments.length==1 ?
							[that, this.n, Math.max(10,this.m)] : arguments));
		if ((typeof that) == 'boolean') return that;

		// now compare the edges using sorted edge lists
		// note: cannot rely on super.equals since directed graphs
		// separate inputs from ouputs in sorted adjacency lists
		let el1 = this.sortedElist(); let el2 = that.sortedElist();
		for (let i = 0; i < el1.length; i++) {
			let e1 = el1[i]; let e2 = el2[i];
			let t1 = this.tail(e1); let t2 = that.tail(e2);
			let h1 = this.head(e1); let h2 = that.head(e2);
			if (t1 != t2 || h1 != h2) return false;
			for (let j = 0; j < this.propNames.length; j++) {
				let ep1 = this.edgeProperties[j][e1];
				let ep2 = that.edgeProperties[j][e2];
				if ((typeof ep1 != 'number' && typeof ep1 != 'string') ||
					typeof ep1 != typeof ep2)
					continue;
				if ((typeof ep1 == 'number' && ep1 != ep2) ||
					(typeof ep1 == 'string' && ep1.localeCompare(ep2)))
					return false;
			}
		}
		return that;
	}

	/** Create a string representation of an edge.
	 *  @param e is an edge number
	 *  @return a string representing the edge
	 */
	edge2string(e, label, terse) {
		return e == 0 ? '-' :
				((terse && this.n <= 26) ?
					(this.x2s(this.tail(e), label) + 
					 this.x2s(this.head(e), label)) : 
					('(' + this.x2s(this.tail(e), label) + ','  +
					  this.x2s(this.head(e), label) +
					  (this.length ? ',' + this.length(e) : '') + ')'));
	}

	/** Construct a string representation of the Digraph object.
	 *  Uses elab to modify the behavior of the Graph.toString() method.
	 */
	toString(fmt=0, elab=0, vlab=0) {
		if (!elab) {
			elab = (e,u) =>
					(u == this.head(e) ? '' :
					 (this.x2s(this.mate(u,e)) +
						(this.edgeProperties.length > 0 &&
						 this.edgeProperties[0][e] ?
							(':' + this.edgeProperties[0][e]) : '')));
		}
		return super.toString(fmt, elab, vlab);
	}

	/** Initialize graph from a string representation.
	 *  @param s is a string representing a graph; if s includes
	 *  edge properties, they are assumed to represent lengths
	 *  @return true on success, else false
	 */
	static fromString(s, n=10, erange=n, propName='length', defVal=0) {
		// function to parse a vertex
		let vnext = (sc => {
						let u = sc.nextIndex();
						n = Math.max(n,u);
						return u > 0 ? u : 0;
					});
		// function to parse an adjacency list item and save properties
		let pairs = []; let pvals = []; let hasProps = 0;
		let enext = ((u,sc) => {
						let v = sc.nextIndex();
						if (v < 0) return false;
						n = Math.max(n,v);
						let i = pairs.length;
						pairs.push([u,v]);
						if (sc.verify(':',0)) {
							let w = sc.nextNumber();
							if (Number.isNaN(w)) return false;
							pvals[i] = w;
							hasProps = 1;
						}
						return true;
					});

		let sc = new Scanner(s);
		if (!sc.verify('{')) return null;
		while (!sc.verify('}')) {
			let u = vnext(sc);
			if (!u) return null;
			if (sc.verify('[')) {
				while (!sc.verify(']')) {
					if (!enext(u,sc)) return null;
				}
			}
		}

		let g = new Digraph(n, Math.max(erange, pairs.length));
		if (hasProps) g.addEdgeProperty(propName, defVal);

		// configure graph
		for (let i = 0; i < pairs.length; i++) {
			let [u,v] = pairs[i];
			let e = g.join(u,v);
			if (pvals[i] != undefined) {
				g.edgeProperties[0][e] = pvals[i];
			}
		}
		return g;
	}

	/** Randomize the order of the vertices, edges and adjacency lists.
	 *  @return the permutation used for the edges
	 */
	scramble() {
		let ep = super.scramble();
		// separate inputs/outputs in epLists
		for (let u = 1; u <= this.n; u++) {
			this.firstEpOut[u] = 0;
			let ep = this.firstEp[u];
			if (ep == 0) continue;
			let epl = ep;
			while (ep != 0 && ep != this.firstEpOut[u]) {
				// if ep is an output, move it to end of eplist
				let epNext = this.epLists.next(ep);
				if (ep%2 == 0) {
					epl = this.epLists.delete(ep, epl);
					epl = this.epLists.join(epl, ep);
					if (this.firstEpOut[u] == 0)
						this.firstEpOut[u] = ep;
				}
				ep = epNext;
			}
			this.firstEp[u] = epl;
		}
		return ep;
	}
	
	/** Compute random lengths for all the edges.
	 *  @param rand is a random number generator used to generate the
	 *  random edge lengths; it is invoked using any extra arguments
	 *  provided by caller; for example randomLengths(randomInteger, 1, 10)
     *  will assign random integer lengths in 1..10.
	 */
	randomLengths(rand, ...args) {
		if (!this.length) this.addEdgeProperty('length', 0);
		for (let e = this.first(); e; e = this.next(e)) {
			this.length(e, rand(...args));
		}
	}
}
